In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/Dulce/12-2023 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

# LESSON

## Missing Values

Finding missing values

In [4]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [5]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [6]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Addressing Missing Values

#Replace by Mean
df_prods.describe()
df['column with missings'].fillna(mean value, inplace=True)
df['column with missings'].fillna(median value, inplace=True)

In [7]:
#Filter out
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [8]:
#Drop missing values
df_prods.dropna(subset = ['product_name'], inplace = True)

##  Duplicates

### Finding Duplicates

In [9]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Addressing Duplicates

In [10]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [11]:
df_prods_clean_no_dups.shape

(49672, 5)

## Export and Tiding Up

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# TASK

## Task 1 - Import ords.csv and see shape

In [12]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))
df_ords.shape

(3421083, 8)

## Task 2 -  Table Statistic

In [13]:
# Table for Orders
df_ords.describe()

Unnamed: 0      order_id       user_id      quantity  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

ANSWER 2 for Orders: I do not see any inconsistencies in the table. 
        #Quantity: Min(1) and Max(100)
        #Days of Week: Min(0) and Max(6)
        #Hour of Day: Min(0) and Max(23)
        #Days since Prior: Min(0) and Max(30)

In [14]:
# Table for Products
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

ANSWER 2 for Products: I see inconsistencies in the Prices column because 99999€ seems to be too high. I detected it because the percentiles are 4, 7 and 11. The value 99999 seems to be an outliner. 

## Task 3 & 4 -  Mixed-type data

In [15]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

ANSWER 3-4: No mixed-type data found


## Task 5 & 6 - Missing Values

Finding Missing Values

In [16]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
quantity                       0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Option 1: Imput Mean to days_since_prior_order missing values

df_ords['days_since_prior_order'].fillna(1.111484e+01, inplace=True)

Option 2: Imput Median to days_since_prior_order missing values

df_ords['days_since_prior_order'].median()

df_ords['days_since_prior_order'].fillna(8, inplace=True)

Option 3: Create new Dataset

df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

 Option 4: Dropping out Missing Values

df_ords.dropna(subset = ['days_since_prior_order'], inplace = True)

ANSWER 5-6: We see that the 6% of the rows in 'days_since_prior_order' are missing. It is too much, so I would not delete them. Imputing a mean of 11 or a median of 8 would be a nice option. However, I would go for a Option 3 (create a new Dataset), there is a reason behind those missing values so I will keep them. The first orders do not have "days since prior order" so, we should continue working with them as they are real orders with valuable information. To sum up, I will continue working with the df_ords with missing values in this column. 

## Task 7 & 8 - Check for duplicate values

In [17]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [18]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, quantity, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

ANSWER 7-8: No duplicates were found in df_ords_clean

## Task 9 - Export

In [19]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))